File: go/index_behaviour_test.go
package web

import (
	"github.com/playwright-community/playwright-go"
	"log"
	"tdd-html-templates/todo"
	"testing"
)

func Test_toggleTodoItem(t *testing.T) {
	// render the initial HTML
	model := todo.NewList().
		AddWithId(101, "One").
		AddWithId(102, "Two")
	initialHtml := renderTemplate("index.tmpl", model, "/")

	// open the browser page with Playwright
	page := openPage()
	defer page.Close()
	logActivity(page)

	// stub network calls
	err := page.Route("**", func(route playwright.Route) {
		if route.Request().URL() == "http://localhost:4567/index.html" {
			// serve the initial HTML
			stubResponse(route, initialHtml.String(), "text/html")
		} else {
			// avoid unexpected requests
			panic("unexpected request: " + route.Request().URL())
		}
	})
	if err != nil {
		t.Fatal(err)
	}

	// load initial HTML in the page
	response, err := page.Goto("http://localhost:4567/index.html")
	if err != nil {
		t.Fatal(err)
	}
	if response.Status() != 200 {
		t.Fatalf("unexpected status: %d", response.Status())
	}
}

func stubResponse(route playwright.Route, text string, contentType string) {
	err := route.Fulfill(playwright.RouteFulfillOptions{ContentType: &contentType, Body: text})
	if err != nil {
		panic(err)
	}
}

func logActivity(page playwright.Page) {
	page.OnRequest(func(request playwright.Request) {
		log.Printf(">> %s %s\n", request.Method(), request.URL())
	})
	page.OnResponse(func(response playwright.Response) {
		log.Printf("<< %d %s\n", response.Status(), response.URL())
	})
	page.OnLoad(func(page playwright.Page) {
		log.Println("Loaded: " + page.URL())
	})
	page.OnConsole(func(message playwright.ConsoleMessage) {
		log.Println("!  " + message.Text())
	})
}

func openPage() playwright.Page {
	pw, err := playwright.Run()
	if err != nil {
		log.Fatalf("could not start playwright: %v", err)
	}
	browser, err := pw.Chromium.Launch()
	if err != nil {
		log.Fatalf("could not launch browser: %v", err)
	}
	page, err := browser.NewPage()
	if err != nil {
		log.Fatalf("could not create page: %v", err)
	}
	return page
}

File: java/src/test/java/org/example/IndexBehaviourTest.java
package org.example;

import com.microsoft.playwright.*;
import com.microsoft.playwright.options.AriaRole;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;
import org.junit.jupiter.api.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

import static org.assertj.core.api.Assertions.assertThat;
import static org.example.IndexTemplateTest.parseHtml;
import static org.example.IndexTemplateTest.renderTemplate;
import static org.junit.jupiter.api.Assertions.fail;

public class IndexBehaviourTest {
    static Playwright playwright;
    static Browser browser;

    @BeforeAll
    static void launchBrowser() {
        playwright = Playwright.create();
        browser = playwright.chromium().launch();
    }

    @AfterAll
    static void closeBrowser() {
        playwright.close();
    }

    private void logActivity(Page page) {
        page.onRequest(request -> System.out.printf(">> %s %s%n", request.method(), request.url()));
        page.onResponse(response -> System.out.printf("<< %s %s%n", response.status(), response.url()));
        page.onLoad(page1 -> System.out.println("Loaded: " + page1.url()));
        page.onConsoleMessage(consoleMessage -> System.out.println("!  " + consoleMessage.text()));
    }

    String stubbedHtml = """
            <section class="todoapp">
                <p>Stubbed html</p>
            </section>
            """;

    @Test
    void toggleTodoItem() {
        // Render our initial html
        TodoList model = new TodoList()
                .add(101, "One")
                .add(102, "Two");
        String initialHtml = renderTemplate("/index.tmpl", model, "/");

        try (Page page = browser.newPage()) {
            logActivity(page);

            // stub network calls
            page.route("**", route -> {
                if (route.request().url().equals("http://localhost:4567/index.html")) {
                    // serve the initial HTML
                    route.fulfill(new Route.FulfillOptions()
                            .setContentType("text/html")
                            .setBody(initialHtml));
                } else if (route.request().url().equals("http://localhost:4567/toggle/101") && route.request().method().equals("POST")) {
                    // we expect that a POST /toggle/101 request is made when we click on the "One" checkbox
                    route.fulfill(new Route.FulfillOptions()
                            .setContentType("text/html")
                            .setBody(stubbedHtml));
                } else if (route.request().url().equals("https://unpkg.com/htmx.org@1.9.12")) {
                    // serve the htmx library
                    route.fulfill(new Route.FulfillOptions()
                            .setContentType("text/html")
                            .setBody(readFile("/htmx.min.js")));
                } else {
                    // we don't want unexpected calls
                    fail(String.format("Unexpected request: %s %s", route.request().method(), route.request().url()));
                }
            });

            // load initial html
            page.navigate("http://localhost:4567/index.html");

            // click on the "One" checkbox
            var checkbox = page.getByRole(AriaRole.CHECKBOX, new Page.GetByRoleOptions().setName("One"));
            checkbox.click();

            // check that the page has been updated
            var document = parseHtml(page.content());
            var elements = document.select("body > section.todoapp > p");
            assertThat(elements.text())
                    .describedAs(page.content())
                    .isEqualTo("Stubbed html");
        }
    }

    private String readFile(String fileName) {
        try {
            return Files.readString(Paths.get("src/test/resources", fileName));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}
